packages/middleware-flexible-checksums/src/stringUnionSelector.ts

```diff
@@ -10,18 +10,18 @@ export enum SelectorType {
  *
  * @internal
  */
-export const stringUnionSelector = (
+export const stringUnionSelector = <U extends object, K extends keyof U>(
   obj: Record<string, string | undefined>,
   key: string,
-  union: Record<string, string>,
+  union: U,
   type: SelectorType
-) => {
+): U[K] | undefined => {
   if (!(key in obj)) return undefined;
 
   const value = obj[key]!.toUpperCase();
   if (!Object.values(union).includes(value)) {
     throw new TypeError(`Cannot load ${type} '${key}'. Expected one of ${Object.values(union)}, got '${obj[key]}'.`);
   }
 
-  return value;
+  return value as U[K];
```

Comment:
note to self: the type `U[K]` is guaranteed by the thrown error. `value` must be a value of `union`.